csfieldguide/templates/interactives/clicking-with-shaking.html

```diff
@@ -0,0 +1,116 @@
+        <button type="button" class="btn btn-outline-secondary float-right" id="mute">Mute</button>
+      </div>
+    </div>
+
+    <div id="game-view-main-area" class="d-flex">
+      <div class="row align-self-center w-100 no-gutters">
+        <div class="col d-flex justify-content-center w-100">
+          <button type="button" class="btn-success" id="target" hidden>OK</button>
+          <button type="button" id="play" class="btn btn-primary">Play</button>
+        </div>
+      </div>
+    </div>
+
+    <div class="row w-100 no-gutters mb-2">
+      <div class="col mr-2">
+        <button type="button" class="btn btn-outline-danger float-right" id="give-up">Give Up</button>
+      </div>
+    </div>
+
+    <div class="row w-100 no-gutters mb-2">
+      <div class="col ml-2 mr-2">
+        <div class="progress">
+          <div id="game-progress" class="progress-bar bg-primary progress-bar-striped progress-bar-animated"
+               role="progressbar" style="width: 0" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">
+            0/10
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+
+  <div id="results-view" class="mt-4 mb-4">
+    <table id="results-table" class="table table-striped table-hover table-responsive-sm">
+      <thead class="thead-dark">
+      <tr>
+        <th scope="col">Set</th>
+        <th scope="col">Button size (w x h)</th>
+        <th scope="col">No. hits</th>
+        <th scope="col">No. misses</th>
+        <th scope="col">Accuracy (%)</th>
+        <th scope="col">Total time (ms)</th>
+        <th scope="col">Average time per hit (ms)</th>
+      </tr>
+      </thead>
+      <tbody id="results-table-body"></tbody>
+    </table>
+  </div>
+</div>
+{% endblock html %}
+
+{% block css %}
+<link rel="stylesheet" href="{% static 'interactives/clicking-with-shaking/css/clicking-with-shaking.css' %}">
+<link href="https://cdn.datatables.net/1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
+<link href="https://cdn.datatables.net/buttons/2.0.1/css/buttons.dataTables.min.css" rel="stylesheet">
+{% endblock css %}
+
+{% block js %}
+<script src="//code.jquery.com/ui/1.13.0/jquery-ui.js"></script>
+
+<script>
+  const hitSoundURL = "{% static 'interactives/clicking-with-shaking/sfx/hit.wav' %}";
+  const missSoundURL = "{% static 'interactives/clicking-with-shaking/sfx/miss.ogg' %}";
+</script>
+
+<script type="text/javascript" src="{% static 'interactives/clicking-with-shaking/js/clicking-with-shaking.js' %}">
+</script>
+<script type="text/javascript" src="https://cdn.datatables.net/1.11.3/js/jquery.dataTables.min.js"></script>
+<script type="text/javascript" src="https://cdn.datatables.net/buttons/2.0.1/js/dataTables.buttons.min.js"></script>
+<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jszip/3.1.3/jszip.min.js"></script>
+<script type="text/javascript" src="https://cdn.datatables.net/buttons/2.0.1/js/buttons.html5.min.js"></script>
```

Comment:
There is a lot of external dependencies here, are all of them needed? 

If they are needed, like dataTables, then we should include them via NPM ([link here](https://www.npmjs.com/package/datatables.net)). We intially copied them into our repository but they become outdated easily, using a CDN means it might update without us wanting to. NPM allows us to lock version, and use our dependency management tools to track them.

With the little time you have left, I can update the dependencies, but if you could explain their usages that would save me some time.

Reply:
I believe they are all needed:

- All the datatables ones are for the table at the bottom. Surprisingly, BootStrap tables don't allow the sorting of rows by clicking a column. I tried a plugin for BootStrap but it didn't work very well (I don't remember specifically why). Datatables is a popular option. It also allows for easy exporting as a csv or Excel file, which Tim said would be useful.
- jquery-ui is needed for the animations. I used jquery instead of CSS animations as it allows for greater control over the animation (e.g. introduce randomness in how many times to shake vertically and horizontally).
- jszip is needed for the Datatables Excel export to work correctly.